### PR TITLE
Add a test tagging system and mark some tests important

### PR DESCRIPTION
**Pull request by pitrou**
The "important" subset is roughly 4x faster than the whole test suite. Note the command-line option only works on Python 3.4+.

This PR also takes the opportunity to remove some duplicate or irrelevant tests (for example all tests of type-specific ufunc loops in object mode).

### CODE REVIEW

**Comment by stefanseefeld** [comment]
I like the idea of defining subsuites of tests, in particular in conjunction with more advanced CI use-cases. However, I'm not quite sure that tagging tests themselves for this is the best approach - it doesn't scale well. Wouldn't it be possible to define synthetic test suites that contain the desired tests ? Then, instead of running `runtests.py --tags=...` one would simply invoke `runtests.py my_special_test_suite`.
Either way, I think this feature is useful enough to be part of `numba.testing` (rather than just `numba.tests`), in case we want to use it elsewhere, too.)

**Comment by pitrou** [comment]
I think a separate synthetic test suite would be even more work. In any case, both approaches need some maintenance.
As for putting the feature in `numba.testing`, yes, will do!

File: numba/testing/main.py [hunk 1]
```diff
@@ -177,6 +223,55 @@ def run_tests_real():
...
+            # Test method doesn't have any tags
+            pass
+    return unittest.TestSuite(selected)
+
+
+def _choose_random_tests(tests, ratio, seed):
+    """
+    Choose a given proportion of tests at random.
+    """
+    rnd = random.Random()
```
Comment (sklam) [inline numba/testing/main.py]: `random` not imported

Comment (pitrou) [inline numba/testing/main.py]: Nice catch, thank you. I've also added some tests now.

**Comment by stefanseefeld** [comment]
OK, the general infrastructure looks good. I have no opinion on the individual attribution of the 'important' tag to tests.

**Pull request merged**